File: app/src/main/java/com/shindra/Map/MapLineFragment.java
package com.shindra.Map;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.shindra.ErrorScreen;
import com.shindra.HelperLine;
import com.shindra.LoadingScreen;
import com.shindra.MyViewModel;
import com.shindra.R;
import com.shindra.ctslibrary.apibo.RouteType;
import com.shindra.ctslibrary.bo.Line;
import com.shindra.ctslibrary.bo.Stop;
import com.shindra.arrakis.controls.Poi;
import com.shindra.arrakis.observable.ObservableExtensionKt;
import com.shindra.arrakis.observable.ObservableListener;


import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;

public class MapLineFragment extends com.shindra.arrakis.controls.MapFragment
{

    // Attributes
    private static final String TAG = "MapLineFragment";
    private String mTramLineLetter;
    private LoadingScreen mLoadingScreen;
    private ErrorScreen mErrorScreen;

    // Constructor
    public static MapLineFragment onInstance(String tramLineLetter){
        MapLineFragment fragment = new MapLineFragment();
        Bundle args = new Bundle();
        args.putString("tramLineLetter", tramLineLetter);
        fragment.setArguments(args);
        return fragment;
    }

    // Methods
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        Log.i(TAG, "onCreateView");
        View v = inflater.inflate(R.layout.fragment_gmaps,container, false);

        // Retrieve activity data
        mTramLineLetter = getArguments().getString("tramLineLetter");
        ((AppCompatActivity) getContext()).getSupportActionBar().setTitle(getString(R.string.line)+ " " + mTramLineLetter);
        return v;
    }

    @Override
    public void onStart() {
        super.onStart();

        mLoadingScreen = new LoadingScreen(getContext());
        mErrorScreen = new ErrorScreen(getContext());

        // Network calls
        MyViewModel model = new ViewModelProvider(this).get(MyViewModel.class);
        ObservableExtensionKt.observe(model.lineWithStop(RouteType.TRAM, mTramLineLetter), new ObservableListener<Line>() {
            @Override
            public void onLoading()
            {
                Log.i(TAG, "onLoading");
                mLoadingScreen.show();
            }

            @Override
            public void onSuccess(Line data)
            {
                Log.i(TAG, "onSuccess");
                mLoadingScreen.dismiss();
                UpdateView(data);
            }

            @Override
            public void onError(@NotNull Throwable throwable)
            {
                Log.i(TAG, "onError");
                mLoadingScreen.dismiss();
                mErrorScreen.show();
            }
        });
    }

    public void UpdateView(Line line)
    {
        // Add positions on map
        Log.i(TAG, "UpdateView");
        ArrayList<Poi> positions = new ArrayList<Poi>();
        for (Stop element : line.getStops())
            positions.add(new Poi(R.drawable.icon_maps_place_24px, HelperLine.GetLineColor(line.getName()), element.getPosition().getLatitude(), element.getPosition().getLongitude()));
        addPois(positions);
    }
}

File: app/src/main/java/com/shindra/map/MapLineFragment.kt
package com.shindra.map

import android.app.Activity
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.shindra.*
import com.shindra.ErrorScreen
import com.shindra.LoadingScreen
import com.shindra.HelperLine
import com.shindra.arrakis.controls.MapFragment
import com.shindra.arrakis.controls.Poi
import com.shindra.arrakis.observable.ObservableListener
import com.shindra.arrakis.observable.observe
import com.shindra.ctslibrary.apibo.RouteType
import com.shindra.ctslibrary.bo.Line
import java.util.*

/* Kotlin remember:
 *
 * In Kotlin, convention for functions/methods is to start with lower case
 *
 * Using ?. on an object method is called "safe call". You can chain it to multiple calls like:
 * (context as AppCompatActivity?)?.supportActionBar?.title. Each time, it will check if returned
 * value is null or not, without any crash. If a null value is detected, it returns null and after
 * we can handle the exception.
 *
 * A "companion" object is equivalent to "static" in JAVA: onInstance is made companion, because
 * it's a method used only on the class and not on an object of the class.
 *
 */

class MapLineFragment : MapFragment() {

    // Attributes
    private lateinit var mTramLineLetter: String

    // Methods
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // onCreateView is called after onCreate: it retrieves the view

        val view = inflater.inflate(R.layout.fragment_gmaps, container, false)

        // Retrieve activity data
        mTramLineLetter = requireArguments().getString(getString(R.string.TramLineLetterGen)).orEmpty()
        (activity as AppCompatActivity).supportActionBar?.title = "${getString(R.string.line)} $mTramLineLetter"
        return view
    }

    override fun onStart() {
        // onStart is called after onCreate view: it handles the logic part (network calls for us)
        super.onStart()
        val mLoadingScreen = LoadingScreen(context as Activity)
        val mErrorScreen = ErrorScreen(context as Activity)

        // Network calls
        val model = ViewModelProvider(this).get(MyViewModel::class.java)
        model.lineWithStop(RouteType.TRAM, mTramLineLetter).observe(object : ObservableListener<Line> {
                override fun onLoading() {
                    mLoadingScreen.show()
                }

                override fun onSuccess(data: Line) {
                    mLoadingScreen.dismiss()
                    updateView(data)
                }

                override fun onError(throwable: Throwable) {
                    mLoadingScreen.dismiss()
                    mErrorScreen.show()
                }
            })
    }

    // Function
    fun updateView(line: Line) {
        // Add positions on map
        val positions = ArrayList<Poi>()
        line.stops?.forEach {
            it.position?.longitude?.let { it1 -> it.position?.latitude?.let { it2 -> Poi(R.drawable.icon_maps_place_24px, HelperLine.getLineColor(mTramLineLetter), it2, it1) } }?.let { it2 -> positions.add(it2) }
        }
        addPois(positions)
    }

    // Constructor
    companion object {
        fun onInstance(tramLineLetter: String, tramLineLetterKey: String): MapLineFragment {
            val fragment = MapLineFragment()
            val args = Bundle()
            args.putString(tramLineLetterKey, tramLineLetter)
            fragment.arguments = args
            return fragment
        }
    }
}